src/types/inc/utils.hpp

```diff
@@ -85,4 +85,33 @@ namespace Microsoft::Console::Utils
     }
 
     GUID CreateV5Uuid(const GUID& namespaceGuid, const gsl::span<const gsl::byte> name);
+
+    // Method Description:
+    // - Base case provided to handle the last argument to CoalesceOptionals<T...>()
+    template<typename T>
+    T CoalesceOptionals(const T& base)
+    {
+        return base;
+    }
+
+    // Method Description:
+    // - Base case provided to throw an assertion if you call CoalesceOptionals(opt, opt, opt)
+    template<typename T>
+    T CoalesceOptionals(const std::optional<T>& base)
+    {
+        static_assert(false, "CoalesceOptionals must be passed a base non-optional value to be used if all optionals are empty");
+        return T{};
+    }
+
+    // Method Description:
+    // - Returns the value from the first populated optional, or a base value if none were populated.
+    template<typename T, typename... Ts>
+    T CoalesceOptionals(const std::optional<T>& t1, Ts&&... t2)
+    {
+        // Initially, I wanted to check "has_value" and short-circuit out so that we didn't
+        // evaluate value_or for every single optional, but has_value/value emits exception handling
+        // code that value_or doesn't. Less exception handling is cheaper than calling value_or a
+        // few more times.

```

Comment:
It was really cool to plug this thing into the Godbolt compiler explorer: if you give `CoalesceOptionals` three optionals whose values are known _at compile time_, the optimizer can actually fully pre-compute the return value.